### PR TITLE
Introducing client_channel_factory

### PR DESCRIPTION
**Pull request by dgquintas**
Former `subchannel_factory`. Added the ability to create channels from client_channel_factory. This
will be used by the grpclb code to create the channels to communicate with the LB servers.

### CODE REVIEW

File: src/core/lib/client_config/client_channel_factory.h [hunk 1]
```diff
@@ -31,36 +31,48 @@
...
-  void (*ref)(grpc_subchannel_factory *factory);
-  void (*unref)(grpc_exec_ctx *exec_ctx, grpc_subchannel_factory *factory);
+struct grpc_client_channel_factory_vtable {
+  void (*ref)(grpc_client_channel_factory *factory);
+  void (*unref)(grpc_exec_ctx *exec_ctx, grpc_client_channel_factory *factory);
   grpc_subchannel *(*create_subchannel)(grpc_exec_ctx *exec_ctx,
-                                        grpc_subchannel_factory *factory,
+                                        grpc_client_channel_factory *factory,
                                         grpc_subchannel_args *args);
+  grpc_channel *(*create_channel)(grpc_exec_ctx *exec_ctx,
```
Comment (ctiller) [inline src/core/lib/client_config/client_channel_factory.h]: suggest adding an enum of what kind of channel to create (can imagine resolver, load-balancing, etc...) - just need load balancing as a value for now.

Comment (dgquintas) [inline src/core/lib/client_config/client_channel_factory.h]: Would you expose that type all the way up to the public API (namely, https://github.com/grpc/grpc/blob/master/include/grpc/grpc.h#L240 and its secure counterpart)?

Comment (dgquintas) [inline src/core/lib/client_config/client_channel_factory.h]: Would you expose that type all the way up to the public API (namely, https://github.com/grpc/grpc/blob/master/include/grpc/grpc.h#L240 and its secure counterpart)?

**Comment by dgquintas** [comment]
Done. PTAL.

**Comment by ctiller** [comment]
LGTM

**Comment by ctiller** [comment]
Needs a merge

**Comment by dgquintas** [comment]
Done.

**Pull request merged**